Scheduling hangs due to duplicate scheduling specification ID

### Checked for duplicates

Yes - I've already checked

### Is this a regression?

Yes - This worked in a previous version

### Version

1.6.0

### Describe the bug

When running scheduling our database got into a state where scheduling would fail with this error in the scheduler_worker log:
```
Exception in thread "main" java.lang.Error: Claiming a scheduling request for specification id 100 returned more than one result row.
```

### Reproduction

Not sure yet how to reproduce. We are rebuilding the database to try again.

### Logs

```shell
[main] INFO com.zaxxer.hikari.HikariDataSource - HikariPool-1 - Starting...
[main] ERROR com.zaxxer.hikari.pool.HikariPool - HikariPool-1 - Exception during pool initialization.
com.impossibl.postgres.jdbc.PGSQLSimpleException: Connection Error: Connection refused: postgres/172.20.0.2:5432
	at com.impossibl.postgres.jdbc.ErrorUtils.makeSQLException(ErrorUtils.java:138)
	at com.impossibl.postgres.jdbc.ConnectionUtil.createConnection(ConnectionUtil.java:326)
	at com.impossibl.postgres.jdbc.AbstractDataSource.createConnection(AbstractDataSource.java:125)
	at com.impossibl.postgres.jdbc.PGDataSource.getConnection(PGDataSource.java:71)
	at com.zaxxer.hikari.pool.PoolBase.newConnection(PoolBase.java:359)
	at com.zaxxer.hikari.pool.PoolBase.newPoolEntry(PoolBase.java:201)
	at com.zaxxer.hikari.pool.HikariPool.createPoolEntry(HikariPool.java:470)
	at com.zaxxer.hikari.pool.HikariPool.checkFailFast(HikariPool.java:561)
	at com.zaxxer.hikari.pool.HikariPool.<init>(HikariPool.java:100)
	at com.zaxxer.hikari.HikariDataSource.<init>(HikariDataSource.java:81)
	at gov.nasa.jpl.aerie.scheduler.worker.SchedulerWorkerAppDriver.main(SchedulerWorkerAppDriver.java:58)
Caused by: io.netty.channel.AbstractChannel$AnnotatedConnectException: Connection refused: postgres/172.20.0.2:5432
Caused by: java.net.ConnectException: Connection refused
	at java.base/sun.nio.ch.Net.pollConnect(Native Method)
	at java.base/sun.nio.ch.Net.pollConnectNow(Unknown Source)
	at java.base/sun.nio.ch.SocketChannelImpl.finishConnect(Unknown Source)
	at io.netty.channel.socket.nio.NioSocketChannel.doFinishConnect(NioSocketChannel.java:330)
	at io.netty.channel.nio.AbstractNioChannel$AbstractNioUnsafe.finishConnect(AbstractNioChannel.java:334)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:707)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:655)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:581)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:493)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at java.base/java.lang.Thread.run(Unknown Source)
Exception in thread "main" com.zaxxer.hikari.pool.HikariPool$PoolInitializationException: Failed to initialize pool: Connection Error: Connection refused: postgres/172.20.0.2:5432
	at com.zaxxer.hikari.pool.HikariPool.throwPoolInitializationException(HikariPool.java:596)
	at com.zaxxer.hikari.pool.HikariPool.checkFailFast(HikariPool.java:582)
	at com.zaxxer.hikari.pool.HikariPool.<init>(HikariPool.java:100)
	at com.zaxxer.hikari.HikariDataSource.<init>(HikariDataSource.java:81)
	at gov.nasa.jpl.aerie.scheduler.worker.SchedulerWorkerAppDriver.main(SchedulerWorkerAppDriver.java:58)
Caused by: com.impossibl.postgres.jdbc.PGSQLSimpleException: Connection Error: Connection refused: postgres/172.20.0.2:5432
	at com.impossibl.postgres.jdbc.ErrorUtils.makeSQLException(ErrorUtils.java:138)
	at com.impossibl.postgres.jdbc.ConnectionUtil.createConnection(ConnectionUtil.java:326)
	at com.impossibl.postgres.jdbc.AbstractDataSource.createConnection(AbstractDataSource.java:125)
	at com.impossibl.postgres.jdbc.PGDataSource.getConnection(PGDataSource.java:71)
	at com.zaxxer.hikari.pool.PoolBase.newConnection(PoolBase.java:359)
	at com.zaxxer.hikari.pool.PoolBase.newPoolEntry(PoolBase.java:201)
	at com.zaxxer.hikari.pool.HikariPool.createPoolEntry(HikariPool.java:470)
	at com.zaxxer.hikari.pool.HikariPool.checkFailFast(HikariPool.java:561)
	... 3 more
Caused by: io.netty.channel.AbstractChannel$AnnotatedConnectException: Connection refused: postgres/172.20.0.2:5432
Caused by: java.net.ConnectException: Connection refused
	at java.base/sun.nio.ch.Net.pollConnect(Native Method)
	at java.base/sun.nio.ch.Net.pollConnectNow(Unknown Source)
	at java.base/sun.nio.ch.SocketChannelImpl.finishConnect(Unknown Source)
	at io.netty.channel.socket.nio.NioSocketChannel.doFinishConnect(NioSocketChannel.java:330)
	at io.netty.channel.nio.AbstractNioChannel$AbstractNioUnsafe.finishConnect(AbstractNioChannel.java:334)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:707)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:655)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:581)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:493)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at java.base/java.lang.Thread.run(Unknown Source)
[main] INFO com.zaxxer.hikari.HikariDataSource - HikariPool-1 - Starting...
[main] INFO com.zaxxer.hikari.pool.HikariPool - HikariPool-1 - Added connection com.impossibl.postgres.jdbc.PGDirectConnection@5876a9af
[main] INFO com.zaxxer.hikari.HikariDataSource - HikariPool-1 - Start completed.
[main] INFO io.javalin.Javalin - Starting Javalin ...
[main] INFO org.eclipse.jetty.server.Server - jetty-11.0.12; built: 2022-09-14T02:38:00.723Z; git: d5b8c29485f5f56a14be5f20c2ccce81b93c5555; jvm 19.0.2+7
[main] INFO org.eclipse.jetty.server.session.DefaultSessionIdManager - Session workerName=node0
[main] INFO org.eclipse.jetty.server.handler.ContextHandler - Started i.j.j.@4d18aa28{/,null,AVAILABLE}
[main] INFO org.eclipse.jetty.server.AbstractConnector - Started ServerConnector@71529963{HTTP/1.1, (http/1.1)}{0.0.0.0:8080}
[main] INFO org.eclipse.jetty.server.Server - Started Server@1e636ea3{STARTING}[11.0.12,sto=0] @1614ms
[main] INFO io.javalin.Javalin -
       __                  ___          ______
      / /___ __   ______ _/ (_)___     / ____/
 __  / / __ `/ | / / __ `/ / / __ \   /___ \
/ /_/ / /_/ /| |/ / /_/ / / / / / /  ____/ /
\____/\__,_/ |___/\__,_/_/_/_/ /_/  /_____/

       https://javalin.io/documentation

[main] INFO io.javalin.Javalin - Listening on http://localhost:8080/
[main] INFO io.javalin.Javalin - You are running Javalin 5.1.2 (released October 24, 2022. Your Javalin version is 226 days old. Consider checking for a newer version.).
[main] INFO io.javalin.Javalin - Javalin started in 253ms \o/
[PG-JDBC I/O (1)] INFO gov.nasa.jpl.aerie.scheduler.worker.ListenSchedulerCapability - Received PSQL Notification: 470, scheduling_request_notification, {"specification_revision":33,"specification_id":100,"analysis_id":29}
Exception in thread "main" java.lang.Error: Claiming a scheduling request for specification id 100 returned more than one result row.
	at gov.nasa.jpl.aerie.scheduler.server.remotes.postgres.PostgresResultsCellRepository.claim(PostgresResultsCellRepository.java:77)
	at gov.nasa.jpl.aerie.scheduler.worker.SchedulerWorkerAppDriver.main(SchedulerWorkerAppDriver.java:85)
[PG-JDBC I/O (1)] INFO gov.nasa.jpl.aerie.scheduler.worker.ListenSchedulerCapability - Received PSQL Notification: 470, scheduling_request_notification, {"specification_revision":4,"specification_id":108,"analysis_id":30}
[PG-JDBC I/O (1)] INFO gov.nasa.jpl.aerie.scheduler.worker.ListenSchedulerCapability - Received PSQL Notification: 691, scheduling_request_notification, {"specification_revision":5,"specification_id":108,"analysis_id":31}
```


### System Info

```shell
Chrome, MacOS
```


### Severity

Moderate

--- a/deployment/hasura/migrations/AerieScheduler/10_cancel_scheduling_trigger/down.sql
+++ b/deployment/hasura/migrations/AerieScheduler/10_cancel_scheduling_trigger/down.sql
@@ -1,0 +1,4 @@
+drop trigger cancel_pending_scheduling_rqs on scheduling_request;
+drop function cancel_pending_scheduling_rqs();
+
+call migrations.mark_migration_rolled_back('10');

--- a/deployment/hasura/migrations/AerieScheduler/10_cancel_scheduling_trigger/up.sql
+++ b/deployment/hasura/migrations/AerieScheduler/10_cancel_scheduling_trigger/up.sql
@@ -1,0 +1,19 @@
+create function cancel_pending_scheduling_rqs()
+returns trigger
+security definer
+language plpgsql as $$
+begin
+  update scheduling_request
+  set canceled = true
+  where status = 'pending'
+  and specification_id = new.specification_id;
+  return new;
+end
+$$;
+
+create trigger cancel_pending_scheduling_rqs
+  before insert on scheduling_request
+  for each row
+  execute function cancel_pending_scheduling_rqs();
+
+call migrations.mark_migration_applied('10');

--- a/scheduler-server/sql/scheduler/applied_migrations.sql
+++ b/scheduler-server/sql/scheduler/applied_migrations.sql
@@ -12,3 +12,4 @@ call migrations.mark_migration_applied('6');
 call migrations.mark_migration_applied('7');
 call migrations.mark_migration_applied('8');
 call migrations.mark_migration_applied('9');
+call migrations.mark_migration_applied('10');

--- a/scheduler-server/sql/scheduler/tables/scheduling_request.sql
+++ b/scheduler-server/sql/scheduler/tables/scheduling_request.sql
@@ -44,7 +44,7 @@ comment on column scheduling_request.requested_at is e''
 -- Scheduling request NOTIFY triggers
 -- These triggers NOTIFY LISTEN(ing) scheduler worker clients of pending scheduling requests
 
-create or replace function notify_scheduler_workers ()
+create function notify_scheduler_workers ()
 returns trigger
 security definer
 language plpgsql as $$
@@ -64,9 +64,25 @@ begin
   return null;
 end$$;
 
-do $$ begin
 create trigger notify_scheduler_workers
   after insert on scheduling_request
   for each row
   execute function notify_scheduler_workers();
-end $$;
+
+create function cancel_pending_scheduling_rqs()
+returns trigger
+security definer
+language plpgsql as $$
+begin
+  update scheduling_request
+  set canceled = true
+  where status = 'pending'
+  and specification_id = new.specification_id;
+  return new;
+end
+$$;
+
+create trigger cancel_pending_scheduling_rqs
+  before insert on scheduling_request
+  for each row
+  execute function cancel_pending_scheduling_rqs();

